File: day_six/day_six.exs
# TODO: Refactor!! While both parts one and two return the correct answer, the perfomance is just horrendous.
# solve_part_two completion took roughly 120.3 seconds.
# Potential solution ideas:
#   1. Attempt it with single dimension array?
#   2. Attempt it using an Agent to create mutable state variable to avoid having to loop through all 1mil points.
#     2a. In line with the point above, will attempt this puzzle with Golang to compare. Even aside from baked in perfomance, number of loops could be seriously reduced.
#   3. Also challenge here, aside from sheer volume, is that instructions need to be run sync to ensure correct state changes from begin to end.

defmodule DaySix do
  def read_file do
    result = File.read("./input.txt")

    case result do
      {:ok, content} -> content
      {:error, :enoent} -> IO.puts("error reading input file")
    end
  end

  def create_matrix do
    0..999
    |> Enum.to_list()
    |> Enum.map(fn _i ->
      Enum.to_list(0..999) |> Enum.map(fn _i -> 0 end)
    end)
  end

  def parse_coordinates(parts, pos) do
    parts
    |> Enum.at(pos)
    |> String.split(",")
    |> Enum.map(fn num_string ->
      case Integer.parse(num_string) do
        {num, _} -> num
        _ -> IO.puts("failed to parse string #{num_string}")
      end
    end)
    |> List.to_tuple()
  end

  def parse_instruction(instruction) do
    # turn on 489,959 through 759,964
    parts = String.split(instruction, " ")

    cond do
      String.contains?(instruction, "turn on") ->
        %{change: "turn on", start: parse_coordinates(parts, 2), end: parse_coordinates(parts, 4)}

      String.contains?(instruction, "turn off") ->
        %{
          change: "turn off",
          start: parse_coordinates(parts, 2),
          end: parse_coordinates(parts, 4)
        }

      String.contains?(instruction, "toggle") ->
        %{change: "toggle", start: parse_coordinates(parts, 1), end: parse_coordinates(parts, 3)}
    end
  end

  def handle_turn_on(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction

    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          1
        else
          elem
        end
      end)
    end)
  end

  def handle_turn_off(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction
    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          0
        else
          elem
        end
      end)
    end)
  end

  def handle_toggle(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction
    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          if elem === 1,
          do: 0,
          else: 1
        else
          elem
        end
      end)
    end)
  end

  # Answer is 569999
  def solve_part_one do
    read_file()
    |> String.trim()
    |> String.split("\n")
    |> Enum.reduce(create_matrix(), fn instruction, matrix ->
      parsed_instruction = parse_instruction(instruction)

      case parsed_instruction do
        %{change: "turn on"} -> handle_turn_on(parsed_instruction, matrix)
        %{change: "turn off"} -> handle_turn_off(parsed_instruction, matrix)
        %{change: "toggle"} -> handle_toggle(parsed_instruction, matrix)
      end
    end)
    |> Enum.map(&Enum.sum(&1))
    |> Enum.sum()
    |> IO.inspect()
  end

  def handle_turn_on_two(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction

    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          elem + 1
        else
          elem
        end
      end)
    end)
  end

  def handle_turn_off_two(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction
    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          if elem === 0,
          do: elem,
          else: elem - 1
        else
          elem
        end
      end)
    end)
  end

  def handle_toggle_two(parsed_instruction, matrix) do
    %{start: {start_x, start_y}, end: {end_x, end_y}} = parsed_instruction
    matrix
    |> Enum.with_index()
    |> Enum.map(fn {row, i} ->
      row
      |> Enum.with_index()
      |> Enum.map(fn {elem, j} ->
        if i in (start_y..end_y) && j in (start_x..end_x)  do
          elem + 2
        else
          elem
        end
      end)
    end)
  end

  # Answer is 17836115
  def solve_part_two do
    read_file()
    |> String.trim()
    |> String.split("\n")
    |> Enum.reduce(create_matrix(), fn instruction, matrix ->
      parsed_instruction = parse_instruction(instruction)

      case parsed_instruction do
        %{change: "turn on"} -> handle_turn_on_two(parsed_instruction, matrix)
        %{change: "turn off"} -> handle_turn_off_two(parsed_instruction, matrix)
        %{change: "toggle"} -> handle_toggle_two(parsed_instruction, matrix)
      end
    end)
    |> Enum.map(&Enum.sum(&1))
    |> Enum.sum()
    |> IO.inspect(label: "answer to part two:")
  end
end

defmodule Benchmark do
  def measure(function) do
    function
    |> :timer.tc
    |> elem(0)
    |> Kernel./(1_000_000)
    |> IO.inspect(label: "time to complete:")
  end
end

Benchmark.measure(fn  -> DaySix.solve_part_two() end)

File: day_six/day_six.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func createMatrix() [][]int {
	var matrix [][]int

	for i := 0; i <= 999; i++ {
		row := make([]int, 999)
		for j := 0; j <= 999; j++ {
			row = append(row, 0)
		}
		matrix = append(matrix, row)
	}

	return matrix
}

type Instruction struct {
	Change string
	StartX int
	StartY int
	EndX   int
	EndY   int
}

func parseInstructionLine(i string) Instruction {
	parts := strings.Split(i, " ")
	if len(parts) == 5 {
		start := strings.Split(parts[2], ",")
		end := strings.Split(parts[4], ",")

		startX, _ := strconv.Atoi(start[0])
		startY, _ := strconv.Atoi(start[1])
		endX, _ := strconv.Atoi(end[0])
		endY, _ := strconv.Atoi(end[1])

		return Instruction{
			Change: fmt.Sprintf("%s %s", parts[0], parts[1]),
			StartX: startX,
			StartY: startY,
			EndX:   endX,
			EndY:   endY,
		}
	} else {
		start := strings.Split(parts[1], ",")
		end := strings.Split(parts[3], ",")

		startX, _ := strconv.Atoi(start[0])
		startY, _ := strconv.Atoi(start[1])
		endX, _ := strconv.Atoi(end[0])
		endY, _ := strconv.Atoi(end[1])

		return Instruction{
			Change: parts[0],
			StartX: startX,
			StartY: startY,
			EndX:   endX,
			EndY:   endY,
		}
	}
}

func solvePartOne(instructions []string) int {
	matrix := createMatrix()
	for _, d := range instructions {
		if d == "" {
			continue
		}
		instruction := parseInstructionLine(d)
		for i := instruction.StartY; i <= instruction.EndY; i++ {
			for j := instruction.StartX; j <= instruction.EndX; j++ {
				if instruction.Change == "turn on" {
					matrix[i][j] = 1
				} else if instruction.Change == "turn off" {
					matrix[i][j] = 0
				} else if instruction.Change == "toggle" {
					if matrix[i][j] == 1 {
						matrix[i][j] = 0
					} else {
						matrix[i][j] = 1
					}
				}
			}
		}
	}

	result := 0
	for _, row := range matrix {
		for _, i := range row {
			result += i
		}
	}

	return result
}

func solvePartTwo(instructions []string) int {
	matrix := createMatrix()
	for _, d := range instructions {
		if d == "" {
			continue
		}
		instruction := parseInstructionLine(d)
		for i := instruction.StartY; i <= instruction.EndY; i++ {
			for j := instruction.StartX; j <= instruction.EndX; j++ {
				if instruction.Change == "turn on" {
					matrix[i][j] = matrix[i][j] + 1
				} else if instruction.Change == "turn off" {
					if matrix[i][j] != 0 {
						matrix[i][j] = matrix[i][j] - 1
					}
				} else if instruction.Change == "toggle" {
					matrix[i][j] = matrix[i][j] + 2

				}
			}
		}
	}

	result := 0
	for _, row := range matrix {
		for _, i := range row {
			result += i
		}
	}

	return result
}

func main() {
	data, err := os.ReadFile("./input.txt")
	if err != nil {
		log.Fatal("error reading input")
	}
	instructions := strings.Split(string(data), "\n")
	resultOne := solvePartOne(instructions)
	resultTwo := solvePartTwo(instructions)

	fmt.Printf("Part One Result = %v \n", resultOne)
	fmt.Printf("Part One Expected = %v \n", 569999)
	fmt.Printf("Part Two Result = %v \n", resultTwo)
	fmt.Printf("Part Two Expected = %v \n", 17836115)
}
